src/lib/dnssd/Resolver_ImplMinimalMdns.cpp

```diff
@@ -594,5 +595,23 @@ Resolver & chip::Dnssd::Resolver::Instance()
     return gResolver;
 }
 
+CHIP_ERROR ResolverProxy::ResolveNodeId(const PeerId & peerId, Inet::IPAddressType type, Resolver::CacheBypass dnssdCacheBypass)
+{
+    chip::Dnssd::Resolver::Instance().SetResolverDelegate(this);
```

Comment:
> 

This is correct. `MinimalMdns` does not support multiple delegates internally afaict. Once https://github.com/project-chip/connectedhomeip/blob/master/src/lib/dnssd/minimal_mdns/Server.cpp#L408 receive an udp packet it just forwarded to the registered delegate.

Things are different for the platform implementation as it pass a `void * context` to the underlying stack, and this is the role of the stack to carry over the context during the request lifecycle.

I did this patch assuming `Minimal` is not designed to be used by multiple consumers at the same time. Otherwise it will definitively needs a rewrite...

Reply:
So what is the advantage of having the proxy between the singleton and the caller if you can only have one resolver going at a time? Is it just a filter?

Reply:
I have similar concerns. Maybe `ResolverDelegate*` could instead be passed to `Resolver::ResolveNodeId` in which case only that delegate would be notified of a success. The minimal mDNS-based resolver could store that delegate in `ActiveResolveAttempts` while the platform resolver could pass it as a context in the call to `ChipDnssdResolveNode`.

Reply:
> So what is the advantage of having the proxy between the singleton and the caller if you can only have one resolver going at a time? Is it just a filter?

This is mostly for the use case where you want multiple controllers at the same time. And usually this use case does not involve the `minimal` implementation but some platform implementation. Hence in platform the proxy is dispatching the responses to the appropriate controller.

So the advantage is that it does not restrict us to the limitation of minimal which does not supports multiple consumers at the same time (looking at the code it is unclear to me that is even support doing a browse and a node id resolution at the same type since `mDiscoveryType` will be set to whatever the last method call is)

So minimal is using the same "abstraction" to access the stack while it does not provide much benefits to it.
It does provides some benefits, like  the fact that you can actually use multiple consumers if your calls are serialised without having to manually call `SetResolverDelegate`. And this is enough for me to write some tests for multi-admins.

Reply:
> I have similar concerns. Maybe `ResolverDelegate*` could instead be passed to `Resolver::ResolveNodeId` in which case only that delegate would be notified of a success. The minimal mDNS-based resolver could store that delegate in `ActiveResolveAttempts` while the platform resolver could pass it as a context in the call to `ChipDnssdResolveNode`.

Indeed that would be better. 

I do only have a basic knowledge of the minimal code, but minimal code looks designed as there is a single request type/consumer at the same time and it makes it a bunch of work to fix, when at the same time it is unclear that minimal will ever be used with multiple controllers instance in the same process in the real world.

Furthermore, one of the issue with passing directly a `ResolverDelegate *` is that it can be a dangling pointer when the dns response comes back. And so we will then need code in minimal but also in every platform implementation to track that.

That just seems much more complex to maintain and to implement than adding a "proxy" in-between.

Reply:
Does this actually solve the problem of multiple controllers though? This still seems racy to me because even if each controller has its own proxy, the proxies are still racing when they set themselves as the resolver delegate on the singleton.

Is the goal here to allow multiple controllers to enroll themselves as a delegate and only receive back notifications for resolve attempts that they've sent out? The controller already does have a filter to determine if there's a pending resolve. Would it be possible to solve this issue by making that filter more robust?

Reply:
OK, we sync'd offline and the peice I'm concerned about only relates to minimal, which has other problems for multi-admin anyway. Vivien said he would add a comment about that to stop others from having the same misunderstanding I did. Approving this with that change.